---
title: "R Notebook"
output: html_notebook
---

```{r}
suppressMessages(library("ggplot2"))
suppressMessages(library("ggnewscale"))
suppressMessages(library("gridExtra"))
suppressMessages(library("reshape2"))
suppressMessages(library("qtl"))
suppressMessages(library("tidyr"))
suppressMessages(library("dplyr"))
suppressMessages(library("ggbeeswarm"))
suppressMessages(library("RColorBrewer"))
suppressMessages(library("ggrepel"))
suppressMessages(library("nVennR"))
suppressMessages(library("VennDiagram"))
suppressMessages(library("GenomicRanges"))
suppressMessages(library("ggbio"))
```

Function to report the row counts within some sliding window and step size (function)
```{r}
# df should have chromosome/position columns; chrlen should be a dataframe
sliding_win <- function(df, win_size, sliding) {
    bundle_list <- list()
    for (i in seq(from = 0, to = 0.99, by = sliding)) {
      SNP_dis <- df %>% group_by(chromosome) %>% mutate(bin = cut(position, label = F, breaks = c(seq(i*win_size, max(position), win_size), max(position)))) %>% group_by(chromosome, bin) %>% dplyr::summarise(n = n())
      SNP_dis <- as.data.frame(SNP_dis)
      SNP_dis$bin_start <- (SNP_dis$bin-1)*win_size + i*win_size
      SNP_dis$bin_end <- SNP_dis$bin*win_size + i*win_size
      bundle_list[[as.character(i)]] <- SNP_dis
    }
    bundle_freq <- do.call(rbind, bundle_list)
    bundle_freq <- bundle_freq[with(bundle_freq, order(chromosome, bin_start, bin_end)), ]
    return(bundle_freq)
}
```

classify cis and trans eQTLs based on distance of 0.8 Mb (file)
```{r}
eQTL_df <- read.table("output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak.txt", sep = "\t", header = T)
bin_pos <- read.table("data/bin_position.txt", sep = "\t", header = T, row.names = 1)
colnames(bin_pos) <- c("bin_chromosome", "bin_start", "bin_end", "bin_position")
gene_pos <- read.table("data/gene_loc_2022.txt", sep = "\t", header = T)
gene_pos <- gene_pos[, c("gene", "chromosome", "start", "end", "strand")]
colnames(gene_pos) <- c("gene", "gene_chromosome", "gene_start", "gene_end", "strand")
# add bin and gene position
eQTL_df <- merge(eQTL_df, bin_pos, by.x = "SNP", by.y = 0, all.x = T)
eQTL_df <- merge(eQTL_df, gene_pos, by = "gene", all.x = T)
# add distance for gene and eQTL
eQTL_df$distance <- ifelse(eQTL_df$bin_chromosome == eQTL_df$gene_chromosome, ifelse(eQTL_df$strand == "+", eQTL_df$bin_position - eQTL_df$gene_start, eQTL_df$bin_position - eQTL_df$gene_end), "-")
# separate genes on scaffold (all label as eQTL, no description)
eQTL_pos <- eQTL_df[grepl("chromosome", eQTL_df$gene_chromosome), ]
eQTL_pos_scaffold <- eQTL_df[!grepl("chromosome", eQTL_df$gene_chromosome), ]
# when bin range overlapped with gene region, it is always cis-eQTL
cis_distance <- 8e+5
label <- c()
for (i in seq(1, nrow(eQTL_pos))) {
  dist <- eQTL_pos[i, "distance"]
  if (dist != "-") {
    bstart <- eQTL_pos[i, "bin_start"]
    bend <- eQTL_pos[i, "bin_end"]
    gstart <- eQTL_pos[i, "gene_start"]
    gend <- eQTL_pos[i, "gene_end"]
    overlap <- intersect(seq(bstart, bend), seq(gstart, gend))
    if (length(overlap) > 0 | abs(as.numeric(dist)) < cis_distance) label <- c(label, "cis")
    else label <- c(label, "trans")
  } else label <- c(label, "trans")
}
eQTL_pos$label <- label
eQTL_pos_scaffold$label <- "eQTL"
eQTL_all <- rbind(eQTL_pos, eQTL_pos_scaffold)
detox <- read.table("../data_important/detox_RC.txt", sep = "\t", header = T, comment.char = "#")
eQTL_all$gene_label <- ifelse(eQTL_all$gene %in% detox$gene, "detox", "non-detox")
write.table(eQTL_all, file = "output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans.txt", sep = "\t", quote = F, row.names = F)
```

```{r}
eQTL_df <- read.table("output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans.txt", sep = "\t", header = T)
cis_eQTL <- eQTL_df[eQTL_df$label == "cis", ]
write.table(cis_eQTL[, c("gene", "SNP", "FDR", "distance", "beta")], file = "output/cis_eQTL.txt", sep = "\t", quote = F, row.names = F)
```

number of genes trans-regulated by each window of eQTL (segregated by + and - effects, file)
```{r}
eQTL_df <- read.table("output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans.txt", sep = "\t", header = T)
eQTL_trans_plus <- eQTL_df[eQTL_df$label == "trans" & eQTL_df$beta > 0, ]
eQTL_trans_minus <- eQTL_df[eQTL_df$label == "trans" & eQTL_df$beta < 0, ]
win_df <- function(eQTL_trans) {
  eQTL_trans <- eQTL_trans[, c("bin_chromosome", "bin_position")]
  colnames(eQTL_trans) <- c("chromosome", "position")
  trans_freq <- sliding_win(eQTL_trans, 200000, 1)
}
trans_plus <- win_df(eQTL_trans_plus)
trans_minus <- win_df(eQTL_trans_minus)
trans_n <- merge(trans_plus, trans_minus, by = c("chromosome", "bin", "bin_start", "bin_end"), 
suffixes = c("plus", "minus"), all = T)
trans_n[is.na(trans_n)] <- 0
trans_n$total <- trans_n$nplus + trans_n$nminus
write.table(trans_n, file = "output/trans_eQTL_200k_freq.txt", sep = "\t", quote = F, row.names = F)
```

number of genes trans-regulated by each window of eQTL (plot)
```{r}
trans_n <- read.table("output/trans_eQTL_200k_freq.txt", sep = "\t", header = T)
trans_freq_p <- ggplot(data = trans_n) + theme(panel.background = element_rect(color = "grey", fill = "transparent"), legend.position = "none", panel.grid = element_blank(), axis.text = element_blank()) + geom_hline(yintercept = 100, linetype = "dashed", color = "grey", size = 0.2) + facet_wrap(~chromosome, nrow = 1, scales = "fixed") + geom_histogram(aes(x = bin_start, y = total), fill = "grey60", stat = "identity") + geom_histogram(aes(x = bin_start, y = nplus), fill = "black", stat = "identity") + theme(panel.spacing = unit(0, "cm")) + scale_x_continuous(breaks = seq(0, max(beta$win_end), 5000000)) + xlab("eQTL position (Mb)") + ylab("Gene number (200k)")
ggsave(plot = trans_freq_p, filename = "plot/trans_n.pdf", width = 6, height = 2)
```

trans-eQTL hotspots with regulated gene number > 100 (9 hotspots in total)
```{r}
eQTL_df <- read.table("output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans.txt", sep = "\t", header = T)
eQTL_trans <- eQTL_df[eQTL_df$label == "trans", ]
eQTL_hs1 <- eQTL_trans[eQTL_trans$bin_chromosome == "chromosome_1" & ((eQTL_trans$bin_start >= 12.4e+6 & eQTL_trans$bin_start <= 12.6e+6) | (eQTL_trans$bin_end >= 12.4e+6 & eQTL_trans$bin_end <= 12.6e+6)), ]
eQTL_hs2 <- eQTL_trans[eQTL_trans$bin_chromosome == "chromosome_1" & ((eQTL_trans$bin_start >= 31.4e+6 & eQTL_trans$bin_start <= 31.6e+6) | (eQTL_trans$bin_end >= 31.4e+6 & eQTL_trans$bin_end <= 31.6e+6)), ]
eQTL_hs3 <- eQTL_trans[eQTL_trans$bin_chromosome == "chromosome_2" & ((eQTL_trans$bin_start >= 0.4e+6 & eQTL_trans$bin_start <= 0.6e+6) | (eQTL_trans$bin_end >= 0.4e+6 & eQTL_trans$bin_end <= 0.6e+6)), ]
eQTL_hs4 <- eQTL_trans[eQTL_trans$bin_chromosome == "chromosome_2" & ((eQTL_trans$bin_start >= 1.4e+6 & eQTL_trans$bin_start <= 1.6e+6) | (eQTL_trans$bin_end >= 1.4e+6 & eQTL_trans$bin_end <= 1.6e+6)), ]
eQTL_hs5 <- eQTL_trans[eQTL_trans$bin_chromosome == "chromosome_2" & ((eQTL_trans$bin_start >= 2.6e+6 & eQTL_trans$bin_start <= 3.2e+6) | (eQTL_trans$bin_end >= 2.6e+6 & eQTL_trans$bin_end <= 3.2e+6)), ]
eQTL_hs6 <- eQTL_trans[eQTL_trans$bin_chromosome == "chromosome_2" & ((eQTL_trans$bin_start >= 11.4e+6 & eQTL_trans$bin_start <= 11.6e+6) | (eQTL_trans$bin_end >= 11.4e+6 & eQTL_trans$bin_end <= 11.6e+6)), ]
eQTL_hs7 <- eQTL_trans[eQTL_trans$bin_chromosome == "chromosome_2" & ((eQTL_trans$bin_start >= 21.4e+6 & eQTL_trans$bin_start <= 21.6e+6) | (eQTL_trans$bin_end >= 21.4e+6 & eQTL_trans$bin_end <= 21.6e+6)), ]
eQTL_hs8 <- eQTL_trans[eQTL_trans$bin_chromosome == "chromosome_3" & ((eQTL_trans$bin_start >= 7.4e+6 & eQTL_trans$bin_start <= 7.6e+6) | (eQTL_trans$bin_end >= 7.4e+6 & eQTL_trans$bin_end <= 7.6e+6)), ]
eQTL_hs9 <- eQTL_trans[eQTL_trans$bin_chromosome == "chromosome_3" & ((eQTL_trans$bin_start >= 14.0e+6 & eQTL_trans$bin_start <= 14.2e+6) | (eQTL_trans$bin_end >= 14.0e+6 & eQTL_trans$bin_end <= 14.2e+6)), ]
write_gene <- function(df, out) {
  detox_df <- df[df$gene_label == "detox", ]
  detox_hs <- detox_df[, c("gene", "SNP", "gene_label", "beta", "FDR")]
  df_hs <- df[, c("gene", "SNP", "gene_label", "beta", "FDR")]
  write.table(df_hs, file = out, sep = "\t", quote = F, row.names = F)
  write.table(detox_hs, file = paste0(dirname(out), paste0("/detox.", basename(out))), sep = "\t", quote = F, row.names = F)
  print(paste(nrow(df_hs), nrow(detox_hs)))
}
write_gene(eQTL_hs1, "output/hotspot/hotspot1.txt")
write_gene(eQTL_hs2, "output/hotspot/hotspot2.txt")
write_gene(eQTL_hs3, "output/hotspot/hotspot3.txt")
write_gene(eQTL_hs4, "output/hotspot/hotspot4.txt")
write_gene(eQTL_hs5, "output/hotspot/hotspot5.txt")
write_gene(eQTL_hs6, "output/hotspot/hotspot6.txt")
write_gene(eQTL_hs7, "output/hotspot/hotspot7.txt")
write_gene(eQTL_hs8, "output/hotspot/hotspot8.txt")
write_gene(eQTL_hs9, "output/hotspot/hotspot9.txt")
```

combine all trans-regulated genes by the 9 hotspots (file)
```{r}
files <- list.files(path = "output/hotspot/")
files <- files[!grepl("detox", files)]
df_list <- list()
for (f in files) {
  df <- read.table(file.path("output/hotspot/", f), sep = "\t", header = T)
  pre <- strsplit(f, split = ".txt")[[1]]
  df$HS <- pre
  df_list[[pre]] <- df
}
df_hs <- do.call(rbind, df_list)
detox <- read.table("../data_important/detox_RC.txt", sep = "\t", header = T, comment.char = "#")
detox$label <- paste0(detox$family, "-", detox$name)
gene_loc <- read.table("../data_important/gene_loc_2022_description.txt", sep = "\t", header = T, quote = "")
df_hs <- merge(df_hs, detox[, c("gene", "label")], by = "gene", all.x = T)
df_hs <- merge(df_hs, gene_loc, by = "gene", all.x = T)
df_hs <- df_hs[with(df_hs, order(HS, label, description)), ]
write.table(df_hs, file = "output/HS_trans_gene.txt", sep = "\t", quote = F, row.names = F)
```

directional test of effect sizes for genes regulated by the 9 hotspots
```{r}
hs_c <- c()
chi_p_c <- c()
plus_c <- c()
minus_c <- c()
#
files <- list.files(path = "output/hotspot/")
files <- files[!grepl("detox", files)]
for (f in files) {
  df <- read.table(file.path("output/hotspot/", f), sep = "\t", header = T)
  plus <- nrow(df[df$beta > 0, ])
  minus <- nrow(df[df$beta < 0, ])
  chi_p <- chisq.test(x = c(plus, minus))
  chi_p_c <- c(chi_p_c, chi_p$p.value)
  plus_c <- c(plus_c, plus)
  minus_c <- c(minus_c, minus)
  pre <- strsplit(f, split = ".txt")[[1]]
  hs_c <- c(hs_c, pre)
}
hs_chi <- data.frame(hs = hs_c, plus = plus_c, minus = minus_c, chi_p = chi_p_c)
hs_chi$padj <- p.adjust(hs_chi$chi_p, method = "bonferroni")
hs_chi$sig <- ifelse(hs_chi$padj < 0.01, "Y", "N")
write.table(hs_chi, file = "output/HS_direction_chisquare.txt", sep = "\t", quote = F, row.names = F)
```

```{r}
eQTL_df <- read.table("output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans_expression.txt", sep = "\t", header = T)
eQTL_cis <- eQTL_df[eQTL_df$label == "cis", ] # 5685
eQTL_trans <- eQTL_df[eQTL_df$label == "trans", ] # 10563
eQTL_hs1 <- eQTL_trans[eQTL_trans$bin_chromosome == "chromosome_1" & ((eQTL_trans$bin_start >= 12.4e+6 & eQTL_trans$bin_start <= 12.6e+6) | (eQTL_trans$bin_end >= 12.4e+6 & eQTL_trans$bin_end <= 12.6e+6)), ]
write.table(eQTL_hs1, file = "output/hotspot1_expression.txt", sep = "\t", quote = F, row.names = F)
```

#############################################################################################

Effects of cis- and trans-eQTL and trans-eQTL hotspots 
```{r}
eQTL_df <- read.table("output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans.txt", sep = "\t", header = T)
eQTL_cis <- eQTL_df[eQTL_df$label == "cis", ] # 5685
eQTL_trans <- eQTL_df[eQTL_df$label == "trans", ] # 10563
hs1 <- read.table("output/hotspot/hotspot1.txt", sep = "\t", header = T)
hs2 <- read.table("output/hotspot/hotspot2.txt", sep = "\t", header = T)
hs3 <- read.table("output/hotspot/hotspot3.txt", sep = "\t", header = T)
hs4 <- read.table("output/hotspot/hotspot4.txt", sep = "\t", header = T)
hs5 <- read.table("output/hotspot/hotspot5.txt", sep = "\t", header = T)
hs6 <- read.table("output/hotspot/hotspot6.txt", sep = "\t", header = T)
hs7 <- read.table("output/hotspot/hotspot7.txt", sep = "\t", header = T)
hs8 <- read.table("output/hotspot/hotspot8.txt", sep = "\t", header = T)
hs9 <- read.table("output/hotspot/hotspot9.txt", sep = "\t", header = T)
# 
effect_list <- list()
effect_list[["cis"]] <- eQTL_cis[, c("gene", "beta", "gene_label")]
effect_list[["trans"]] <- eQTL_trans[, c("gene", "beta", "gene_label")]
effect_list[["hs1"]] <- hs1[, c("gene", "beta", "gene_label")]
effect_list[["hs2"]] <- hs2[, c("gene", "beta", "gene_label")]
effect_list[["hs3"]] <- hs3[, c("gene", "beta", "gene_label")]
effect_list[["hs4"]] <- hs4[, c("gene", "beta", "gene_label")]
effect_list[["hs5"]] <- hs5[, c("gene", "beta", "gene_label")]
effect_list[["hs6"]] <- hs6[, c("gene", "beta", "gene_label")]
effect_list[["hs7"]] <- hs7[, c("gene", "beta", "gene_label")]
effect_list[["hs8"]] <- hs8[, c("gene", "beta", "gene_label")]
effect_list[["hs9"]] <- hs9[, c("gene", "beta", "gene_label")]
# 
effect_df <- do.call(rbind, effect_list)
effect_df$label <- gsub("\\.[0-9]*", "", rownames(effect_df))
write.table(effect_df, file = "output/beta_label.txt", sep = "\t", quote = F, row.names = F)
```

p-values of cis/trans for all/detox genes (violin plot)
```{r}
eQTL_df <- read.table("output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans_expression.txt", sep = "\t", header = T)
eQTL_p <- eQTL_df[, c("label", "gene_label", "FDR")]
eQTL_e <- eQTL_df[, c("label", "gene_label", "beta")]
eQTL_p <- eQTL_p[eQTL_p$label %in% c("cis", "trans"), ]
eQTL_e <- eQTL_e[eQTL_e$label %in% c("cis", "trans"), ]
# 
eQTL_p$label <- factor(eQTL_p$label, levels = c("cis", "trans"))
eQTL_p$class <- paste0("all", eQTL_p$label)
detox_p <- eQTL_p[eQTL_p$gene_label == "detox", ]
detox_p$class <- paste0("detox", detox_p$label)
# 
eQTL_e$class <- paste0("all", eQTL_e$label)
detox_e <- eQTL_e[eQTL_e$gene_label == "detox", ]
detox_e$class <- paste0("detox", detox_e$label)
# 
pairwise.wilcox.test(-log10(eQTL_p$FDR), eQTL_p$class, p.adjust.method = "bonf")
pairwise.wilcox.test(-log10(detox_p$FDR), detox_p$class, p.adjust.method = "bonf")
#
pairwise.wilcox.test(abs(eQTL_e$beta), eQTL_e$class, p.adjust.method = "bonf")
pairwise.wilcox.test(abs(detox_e$beta), detox_e$class, p.adjust.method = "bonf")
# 
eQTL_cis <- rbind(eQTL_p[eQTL_p$label == "cis" & eQTL_p$gene_label == "non-detox", ], detox_p[detox_p$label == "cis", ])
eQTL_trans <- rbind(eQTL_p[eQTL_p$label == "trans" & eQTL_p$gene_label == "non-detox", ], detox_p[detox_p$label == "trans", ])
eQTL_w <- rbind(eQTL_cis, eQTL_trans)
#
eQTL_cis %>% group_by(class) %>% summarise(m=mean(-log10(FDR)))
eQTL_trans %>% group_by(class) %>% summarise(m=mean(-log10(FDR)))
# 
pairwise.wilcox.test(-log10(eQTL_cis$FDR), eQTL_cis$class, p.adjust.method = "BH")
pairwise.wilcox.test(-log10(eQTL_trans$FDR), eQTL_trans$class, p.adjust.method = "BH")
# 
pairwise.wilcox.test(-log10(eQTL_w$FDR), eQTL_w$class, p.adjust.method = "BH")
# 
eQTLp_freq <- eQTL_p %>% group_by(class) %>% summarise(n=n()) %>% as.data.frame()
detoxp_freq <- detox_p %>% group_by(class) %>% summarise(n=n()) %>% as.data.frame()
# 
eQTL_pp <- ggplot() + geom_violin(data = eQTL_p, aes(x = class, y = -log10(FDR)), size = 0.1, width = 1.7, fill = "#FFEA46FF", alpha = 0.4) + geom_boxplot(data = eQTL_p, aes(x = class, y = -log10(FDR)), size = 0.1, notch = T, outlier.size = 0.3, width = 0.23, color = "black", alpha = 0.3) + xlab("") + ylab("-log10(FDR)") + theme(panel.background = element_rect(fill = "transparent", color = "grey"), panel.grid = element_blank()) + geom_text_repel(data = eQTLp_freq, aes(x = class, y = 100, label = n)) + ylim(c(0, 150))
detox_pp <- ggplot() + geom_violin(data = detox_p, aes(x = class, y = -log10(FDR)), size = 0.1, width = 1.7, fill = "#30718F", alpha = 0.4) + geom_boxplot(data = detox_p, aes(x = class, y = -log10(FDR)), size = 0.1, notch = T, outlier.size = 0.3, width = 0.23, color = "black", alpha = 0.3) + xlab("") + ylab("-log10(FDR)") + theme(panel.background = element_rect(fill = "transparent", color = "grey"), panel.grid = element_blank()) + geom_text_repel(data = detoxp_freq, aes(x = class, y = 100, label = n)) + ylim(c(0, 150))
# 
ggsave(plot = eQTL_pp, filename = "plot/eQTL_pvalue.pdf", width = 4, height = 4)
ggsave(plot = detox_pp, filename = "plot/detox_pvalue.pdf", width = 4, height = 4)
```

```{r}
bin <- read.table("data/bin_position.txt", sep = "\t", header = T)
bin$length <- bin$bin_end - bin$bin_start
summary(bin$length)
```


collect eQTLs for individual gene (file)
```{r}
eQTL_df <- read.table("output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans.txt", sep = "\t", header = T)
detox <- read.table("../data_important/detox_RC_delete_empty.txt", sep = "\t", header = T, comment.char = "#")
eQTL_df_chr <- eQTL_df[eQTL_df$label %in% c("cis", "trans"), ]
gid <- sort(unique(eQTL_df_chr$gene))
gene_eQTL <- data.frame(gene = gid, eQTL = 0)
for (g in gid) {
  eQTL_g <- eQTL_df_chr[eQTL_df_chr$gene == g, ]
  eQTL_l <- paste(sort(eQTL_g$label), collapse = ",")
  gene_eQTL[gene_eQTL$gene == g, "eQTL"] <- eQTL_l
}
# 
gene_detox <- gene_eQTL[gene_eQTL$gene %in% detox$gene, ]
# 
# write.table(gene_eQTL, file = "gene_eQTL_class.txt", sep = "\t", quote = F, row.names = F)
# write.table(gene_detox, file = "detox_eQTL_class.txt", sep = "\t", quote = F, row.names = F)
# 
gene_detox_freq <- as.data.frame(table(gene_detox$eQTL))
gene_eQTL_freq <- as.data.frame(table(gene_eQTL$eQTL))
colnames(gene_eQTL_freq) <- c("eQTL_class", "freq")
colnames(gene_detox_freq) <- c("eQTL_class", "freq")
gene_eQTL_freq$prop <- round(gene_eQTL_freq$freq/9740*100, 2)
gene_detox_freq$prop <- round(gene_detox_freq$freq/537*100, 2)
write.table(gene_eQTL_freq, file = "output/gene_eQTLs_freq.txt", sep = "\t", quote = F, row.names = F)
write.table(gene_detox_freq, file = "output/gene_detox_freq.txt", sep = "\t", quote = F, row.names = F)
```

```{r}
gene_class <- read.table("gene_eQTL_class.txt", sep = "\t", header = T)
detox_class <- read.table("detox_eQTL_class.txt", sep = "\t", header = T)
gene_class$class <- ifelse(gene_class$eQTL == "cis", "cis", ifelse(gene_class$eQTL == "trans", "trans", ifelse(gene_class$eQTL == "cis,trans", "cis,trans", ifelse(grepl("cis", gene_class$eQTL) & grepl("trans", gene_class$eQTL), "cis,m_trans", ifelse(!grepl("cis", gene_class$eQTL) & grepl("trans", gene_class$eQTL), "m_trans", "other")))))
detox_class$class <- ifelse(detox_class$eQTL == "cis", "cis", ifelse(detox_class$eQTL == "trans", "trans", ifelse(detox_class$eQTL == "cis,trans", "cis,trans", ifelse(grepl("cis", detox_class$eQTL) & grepl("trans", detox_class$eQTL), "cis,m_trans", ifelse(!grepl("cis", detox_class$eQTL) & grepl("trans", detox_class$eQTL), "m_trans", "other")))))
# 
gene_class$gene <- "all"
detox_class$gene <- "detox"
# 
combine_class <- rbind(gene_class, detox_class)
combine_class <- combine_class[, c("gene", "class")]
ptable <- table(combine_class)
chisq <- chisq.test(ptable, simulate.p.value = T, B = 10000)
residuals <- chisq$residuals
rcat <- rownames(ptable)
ccat <- colnames(ptable)
for (r in rcat) {
  for (c in ccat) {
    rclass <- (combine_class[, 1] == r)
    cclass <- (combine_class[, 2] == c)
    ptable[r, c] <- fisher.test(rclass, cclass, alternative = "two.sided")$p.value
  }
}
#
padjtable <- ptable
padj <- p.adjust(ptable, method = "bonf")
i <- 1
for (c in ccat) {
  for (r in rcat) {
    padjtable[r, c] <- padj[i]
    i <- i+1
  }
}
#
print(residuals)
print(ptable)

```

detox gene cis and trans eQTLs 
```{r}
eQTL_df <- read.table("output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans.txt", sep = "\t", header = T)
eQTL_df_chr <- eQTL_df[eQTL_df$label %in% c("cis", "trans"), ]
detox <- read.table("../data_important/detox_RC.txt", sep = "\t", header = T)
# 
detox_eQTL <- eQTL_df_chr[eQTL_df_chr$gene %in% detox$gene, ]
detox_gid <- sort(unique(detox_eQTL$gene))
detox_df <- data.frame(gene = detox_gid, eQTL = 0)
for (g in detox_gid) {
  eQTL_g <- detox_eQTL[detox_eQTL$gene == g, ]
  eQTL_l <- paste(sort(eQTL_g$label), collapse = ",")
  detox_df[detox_df$gene == g, "eQTL"] <- eQTL_l
}
write.table(detox_df, file = "output/detox_eQTLs.txt", sep = "\t", quote = F, row.names = F)
```

frequency of detox gene with cis or/and trans combinations
```{r}
detox_df <- read.table("output/detox_eQTLs.txt", sep = "\t", header = T)
detox_freq <- as.data.frame(table(detox_df$eQTL))
colnames(detox_freq) <- c("eQTL_class", "freq")
write.table(detox_freq, file = "output/detox_eQTL_freq.txt", sep = "\t", quote = F, row.names = F)
```

log2foldchange of genes in response to eQTL genotypes change (file)
```{r}
genotype_df <- read.table("data/eQTL_genotype_all.txt", sep = "\t", header = T, row.names = 1, )
expression_df <- read.table("../data_important/htseq_star_new_v2_DESeq.txt", header = T, row.names = 1)
eQTL_df <- read.table("output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans.txt", sep = "\t", header = T)
eQTL_df_association <- eQTL_df[, c("gene", "SNP")]
# check expression and genotype data match
# identical(colnames(genotype_df), colnames(expression_df))
# go through each association by association
eQTL_df_association$exp0 <- 0
eQTL_df_association$exp1 <- 0
for (a in rownames(eQTL_df_association)) {
  gid <- eQTL_df_association[a, "gene"]
  snp_id <- eQTL_df_association[a, "SNP"]
  genotype_sub <- genotype_df[snp_id, ]
  expression_sub <- expression_df[gid, ]
  geno_exp <- rbind(genotype_sub, expression_sub)
  geno_exp_t <- t(geno_exp)
  colnames(geno_exp_t) <- c("genotype", "expression")
  expression_mean <- geno_exp_t %>% as.data.frame() %>% group_by(genotype) %>% summarise(exp_mean=mean(expression)) %>% as.data.frame()
  eQTL_df_association[eQTL_df_association$gene == gid & eQTL_df_association$SNP == snp_id, "exp0"] <- expression_mean[expression_mean$genotype == 0, "exp_mean"]
  eQTL_df_association[eQTL_df_association$gene == gid & eQTL_df_association$SNP == snp_id, "exp1"] <- expression_mean[expression_mean$genotype == 1, "exp_mean"]
}
eQTL_df <- merge(eQTL_df, eQTL_df_association, by = c("gene", "SNP"), all = T)
write.table(eQTL_df, file = "output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans_expression.txt", sep = "\t", quote = F, row.names = F)
```

test if the biased directionality of effect in each of the cis, trans control (drop)
```{r}
effect_df <- read.table("output/beta_label.txt", sep = "\t", header = T)
detox_effect <- effect_df[effect_df$gene_label == "detox", ]
# 
chisq_direction <- function(effect_df) {
  effect_label <- unique(effect_df$label)
  edf <- data.frame(effect_label = effect_label, positive = 0, negative = 0, p = 0)
  for (e in effect_label) {
    effect_sub <- effect_df[effect_df$label == e, ]
    positive_effect <- effect_sub[effect_sub$beta > 0, ]
    negative_effect <- effect_sub[effect_sub$beta < 0, ]
    chisq_res <- chisq.test(c(nrow(positive_effect), nrow(negative_effect)), simulate.p.value = T, B = 5000)
    c_p <- chisq_res$p.value
    edf[edf$effect_label == e, "positive"] <- nrow(positive_effect)
    edf[edf$effect_label == e, "negative"] <- nrow(negative_effect)
    edf[edf$effect_label == e, "p"] <- c_p
  }
  edf$padj <- p.adjust(edf$p, method = "bonferroni")
  return(edf)
}
#
chisq_direct <- chisq_direction(effect_df)
chisq_direct_detox <- chisq_direction(detox_effect)
#
write.table(chisq_direct, file = "output/chisq.test.txt", sep = "\t", quote = F, row.names = F)
write.table(chisq_direct_detox, file = "output/chisq.test_detox.txt", sep = "\t", quote = F, row.names = F)
```

scatter plot for eQTL and gene associations on genome-wide (plot)
```{r}
eQTL_df <- read.table("output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans.txt", sep = "\t", header = T)
eQTL_df_chr <- eQTL_df[grepl("chromosome", eQTL_df$gene_chromosome), ]
eQTL_df_chr$color_label <- ifelse(eQTL_df_chr$gene_label == "detox" & eQTL_df_chr$label == "cis", "detox_cis", ifelse(eQTL_df_chr$gene_label == "detox" & eQTL_df_chr$label == "trans", "detox_trans", eQTL_df_chr$label))
# order dataframe to control the rank/layer in plot
eQTL_df_chr$color_label <- factor(eQTL_df_chr$color_label, levels = c("detox_cis", "detox_trans", "cis", "trans"))
eQTL_df_chr <- eQTL_df_chr[order(eQTL_df_chr$color_label, decreasing = T), ] # always draw detox gene last (so on the top)
eQTL_df_chr$gene_chromosome <- factor(eQTL_df_chr$gene_chromosome, levels = c("chromosome_3", "chromosome_2", "chromosome_1"))
eQTL_df_chr$bin_chromosome <- factor(eQTL_df_chr$bin_chromosome, levels = c("chromosome_1", "chromosome_2", "chromosome_3"))
# 
eQTL_p <- ggplot(data = eQTL_df_chr, aes(x = bin_position, y = (gene_start+gene_end)%/%2, color = color_label, alpha = color_label, shape = color_label, size = color_label)) + facet_grid(gene_chromosome~bin_chromosome, scales = "fixed") + geom_point(stroke = 0.2) + scale_shape_manual(values = c("cis" = 16, "trans" = 16, "detox_cis" = 19, "detox_trans" = 19)) + scale_color_manual(values = c("cis"= "#FCB036", "trans" = "#23C3E4", "detox_cis" = "#C82803", "detox_trans" = "#3F3994")) + scale_size_manual(values = c("cis" = 0.5, "trans" = 0.5, "detox_cis" = 0.65, "detox_trans" = 0.65)) + scale_alpha_manual(values = c("cis" = 1, "trans" = 1, "detox_cis" = 1, "detox_trans" = 1)) + theme(panel.background = element_rect(color = "grey", fill = "transparent")) + theme(legend.position = "top", legend.background = element_rect(fill = NA, color = "grey"), legend.box.background = element_rect(fill = "white", color = "grey"), axis.text = element_blank(), panel.grid = element_blank()) + theme(aspect.ratio = 1) + guides(color = guide_legend(override.aes = list(alpha = 1))) + scale_x_continuous(breaks = seq(0, max(eQTL_df_chr$bin_position), 5e+6)) + scale_y_continuous(breaks = seq(0, max(eQTL_df_chr$bin_position), 5e+6)) + xlab("eQTL position (Mb)") + ylab("gene position (Mb)") + theme(panel.spacing = unit(0, "cm"))
ggsave(plot = eQTL_p, filename = "plot/FDR0.01_rf0.4_LOD3_chr_peak_CisTrans.pdf", width = 6, height = 6)
```

The significance on genome-wide (plot)
```{r}
eQTL_df <- read.table("output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans.txt", sep = "\t", header = T)
eQTL_df_trans <- eQTL_df[eQTL_df$label == "trans", ]
eQTL_df_trans$color_label <- ifelse(eQTL_df_trans$gene_label == "detox", "detox", eQTL_df_trans$label)
eQTL_df_trans$color_label <- factor(eQTL_df_trans$color_label, levels = c("detox", "trans"))
eQTL_df_trans <- eQTL_df_trans[order(eQTL_df_trans$color_label, decreasing = T), ]
eQTL_df_trans$chromosome <- eQTL_df_trans$bin_chromosome
# 
bg_p <- ggplot() + theme(panel.background = element_rect(color = "grey", fill = "transparent"), legend.position = "none", panel.grid = element_blank(), axis.text = element_blank()) 
trans_sig_p <- bg_p + geom_point(data = eQTL_df_trans, aes(x = bin_position, y = -log10(FDR), color = color_label, shape = color_label, size = color_label)) + facet_wrap(~chromosome, scales = "fixed") + scale_color_manual(values = c( "trans" = "#23C3E4", "detox" = "#3F3994")) + scale_size_manual(values = c("trans" = 0.5, "detox" = 0.5)) + scale_shape_manual(values = c("trans" = 16, "detox" = 19)) + facet_wrap(~chromosome, scales = "fixed") + scale_x_continuous(breaks = seq(0, max(eQTL_df_trans$bin_position), 5000000)) + xlab("") + ylab("-log10(FDR)") + theme(panel.spacing = unit(0, "cm"))
ggsave(plot = trans_sig_p, filename = "plot/FDR0.01_rf0.4_LOD3_chr_peak_Trans_sig.pdf", width = 6, height = 2)
```

effect sizes of cis/trans eQTLs and all 9 hotspots (violin plot)
```{r}
effect_df <- read.table("output/beta_label.txt", sep = "\t", header = T)
effect_df <- effect_df[effect_df$label %in% c("hs1", "hs2", "hs3", "hs4", "hs5", "hs6", "hs7", "hs8", "hs9"), ]
effect_df$label <- factor(effect_df$label, levels = c("hs1", "hs2", "hs3", "hs4", "hs5", "hs6", "hs7", "hs8", "hs9"))
effect_n <- effect_df %>% group_by(label) %>% summarise(n = n()) %>% as.data.frame()
effect_detox <- effect_df[effect_df$gene_label == "detox", ]
effect_detox_n <- effect_detox %>% group_by(label) %>% summarise(n = n()) %>% as.data.frame()
# 
effect_all_p <- ggplot() + geom_hline(yintercept = c(-1, 0, 1), color = "grey", size = 0.2, linetype = "dashed") + geom_violin(data = effect_df, aes(x = label, y = beta), size = 0.1, width = 1.7, fill = "#FFEA46FF", alpha = 0.4) + geom_boxplot(data = effect_df, aes(x = label, y = beta), size = 0.1, notch = T, outlier.size = 0.3, width = 0.23, color = "black", alpha = 0.3) + xlab("") + ylab("effect size") + theme(panel.background = element_rect(fill = "transparent", color = "grey"), panel.grid = element_blank()) + geom_text_repel(data = effect_n, aes(x = label, y = 1, label = n))
# 
effect_detox_p <- ggplot(data = effect_detox, aes(x = label, y = beta)) + geom_hline(yintercept = c(-1, 0, 1), color = "grey", size = 0.2, linetype = "dashed") + geom_violin(size = 0.1, width = 1.7, fill = "#30718F", alpha = 0.4) + geom_boxplot(size = 0.1, notch = T, outlier.size = 0.3, width = 0.23, color = "black", alpha = 0.3) + xlab("") + ylab("effect size") + theme(panel.background = element_rect(fill = "transparent", color = "grey"), panel.grid = element_blank()) + geom_text_repel(data = effect_detox_n, aes(x = label, y = 1, label = n))
pdf("plot/effect_all.pdf", width = 7.8, height = 3)
effect_all_p 
dev.off()
#
pdf("plot/effect_detox.pdf", width = 7.8, height = 3)
effect_detox_p 
dev.off()
```

absolute effect size from cis/trans
```{r}
effect_df <- read.table("output/beta_label.txt", sep = "\t", header = T)
effect_df <- effect_df[effect_df$label %in% c("cis", "trans"), ]
effect_df$label <- factor(effect_df$label, levels = c("cis", "trans"))
effect_n <- effect_df %>% group_by(label) %>% summarise(n = n()) %>% as.data.frame()
effect_detox <- effect_df[effect_df$gene_label == "detox", ]
effect_detox_n <- effect_detox %>% group_by(label) %>% summarise(n = n()) %>% as.data.frame()
# 
effect_all_p <- ggplot() + geom_violin(data = effect_df, aes(x = label, y = abs(beta)), size = 0.1, width = 1.7, fill = "#FFEA46FF", alpha = 0.4) + geom_boxplot(data = effect_df, aes(x = label, y = abs(beta)), size = 0.1, notch = T, outlier.size = 0.3, width = 0.23, color = "black", alpha = 0.3) + xlab("") + ylab("effect size") + theme(panel.background = element_rect(fill = "transparent", color = "grey"), panel.grid = element_blank()) + geom_text_repel(data = effect_n, aes(x = label, y = 1, label = n))
# 
effect_detox_p <- ggplot(data = effect_detox, aes(x = label, y = abs(beta))) + geom_violin(size = 0.1, width = 1.7, fill = "#30718F", alpha = 0.4) + geom_boxplot(size = 0.1, notch = T, outlier.size = 0.3, width = 0.23, color = "black", alpha = 0.3) + xlab("") + ylab("effect size") + theme(panel.background = element_rect(fill = "transparent", color = "grey"), panel.grid = element_blank()) + geom_text_repel(data = effect_detox_n, aes(x = label, y = 1, label = n))
pdf("plot/abs_effect_all.pdf", width = 4, height = 4)
effect_all_p 
dev.off()
#
pdf("plot/abs_effect_detox.pdf", width = 4, height = 4)
effect_detox_p 
dev.off()
```

Genes trans-regulated by HS1 (in a circle plot)
```{r}
eQTL_df <- read.table("output/eQTL_FDR0.01_rf0.4_LOD3_chr_peak_CisTrans_expression.txt", sep = "\t", header = T)
eQTL_df$effect <- ifelse(eQTL_df$beta > 0, paste0(eQTL_df$label, "+"), paste0(eQTL_df$label, "-"))
HS1 <- read.table("output/hotspot1_expression.txt", sep = "\t", header = T)
HS1$effect <- ifelse(HS1$beta > 0, paste0(HS1$label, "+"), paste0(HS1$label, "-"))
chrlen <- read.table("../data_important/chrlen.txt", sep = "\t", header = T)
detox <- read.table("../data_important/detox_RC.txt", sep = "\t", header = T)
#
hs1_gene <- HS1$gene
HS1_eQTL_summ <- data.frame(gene = hs1_gene, eQTLs = 0)
for (h in hs1_gene) {
  eQTL_h <- eQTL_df[eQTL_df$gene == h, ]
  HS1_eQTL_summ[HS1_eQTL_summ$gene == h, "eQTLs"] <- paste(unique(eQTL_h$effect), collapse = ",")
}
HS1_cisplus_gene <- HS1_eQTL_summ[grepl("cis+", HS1_eQTL_summ$eQTLs, fixed = T), ]$gene
HS1_cisminus_gene <- HS1_eQTL_summ[grepl("cis-", HS1_eQTL_summ$eQTLs, fixed = T), ]$gene
#
HS1wcis <- HS1[HS1$gene %in% c(HS1_cisplus_gene, HS1_cisminus_gene), ]
HS1wcis$gene_cis <- ifelse(HS1wcis$gene %in% HS1_cisplus_gene, "cis+", ifelse(HS1wcis$gene %in% HS1_cisminus_gene, "cis-", "."))
write.table(HS1wcis, file = "HS1_w_cis.txt", sep = "\t", quote = F, row.names = F)
#
HS1_cisplus_end1 <- HS1wcis[HS1wcis$gene_cis == "cis+", c("bin_chromosome", "bin_start", "bin_end", "effect")]
HS1_cisplus_end2 <- HS1wcis[HS1wcis$gene_cis == "cis+", c("gene_chromosome", "gene_start", "gene_end", "gene_label")]
HS1_cisminus_end1 <- HS1wcis[HS1wcis$gene_cis == "cis-", c("bin_chromosome", "bin_start", "bin_end", "effect")]
HS1_cisminus_end2 <- HS1wcis[HS1wcis$gene_cis == "cis-", c("gene_chromosome", "gene_start", "gene_end", "gene_label")]
# 
colnames(HS1_cisplus_end1) <- c("chromosome", "start", "end", "effect")
colnames(HS1_cisplus_end2) <- c("chromosome", "start", "end", "gene_label")
colnames(HS1_cisminus_end1) <- c("chromosome", "start", "end", "effect")
colnames(HS1_cisminus_end2) <- c("chromosome", "start", "end", "gene_label")
# 
HS1_cisplus_end1_gr <- makeGRangesFromDataFrame(HS1_cisplus_end1, keep.extra.columns = T, ignore.strand = T)
seqlevels(HS1_cisplus_end1_gr) <- c("chromosome_1", "chromosome_2", "chromosome_3")
seqlengths(HS1_cisplus_end1_gr) <- chrlen[, "length"] 
HS1_cisplus_end2_gr <- makeGRangesFromDataFrame(HS1_cisplus_end2, keep.extra.columns = T, ignore.strand = T)
seqlevels(HS1_cisplus_end2_gr) <- c("chromosome_1", "chromosome_2", "chromosome_3")
seqlengths(HS1_cisplus_end2_gr) <- chrlen[, "length"]
HS1_cisminus_end1_gr <- makeGRangesFromDataFrame(HS1_cisminus_end1, keep.extra.columns = T, ignore.strand = T)
seqlevels(HS1_cisminus_end1_gr) <- c("chromosome_1", "chromosome_2", "chromosome_3")
seqlengths(HS1_cisminus_end1_gr) <- chrlen[, "length"] 
HS1_cisminus_end2_gr <- makeGRangesFromDataFrame(HS1_cisminus_end2, keep.extra.columns = T, ignore.strand = T)
seqlevels(HS1_cisminus_end2_gr) <- c("chromosome_1", "chromosome_2", "chromosome_3")
seqlengths(HS1_cisminus_end2_gr) <- chrlen[, "length"]
# 
HS1_cisplus_end1_gr$to.gr <- HS1_cisplus_end2_gr
HS1_cisminus_end1_gr$to.gr <- HS1_cisminus_end2_gr
#
col1 <- "black"
col2 <- "red"
link_p <- ggplot() + layout_circle(HS1_cisplus_end1_gr, geom = "ideo", fill = NA, alpha = 0.4, radius = 7.5, trackWidth = 0.4, color = "grey") + layout_circle(HS1_cisplus_end1_gr, geom = "ideo", fill = NA, alpha = 0.4, radius = 8.1, trackWidth = 0.4, color = "grey") + layout_circle(HS1_cisplus_end1_gr, geom = "link", linked.to = "to.gr", aes(color = effect), size = 0.4, radius = 8.1, trackWidth = 0.2) + layout_circle(HS1_cisminus_end1_gr, geom = "link", linked.to = "to.gr", aes(color = effect), size = 0.4, radius = 7.5, trackWidth = 0.2) + scale_color_manual(values = c("trans+" = col1, "trans-" = col2)) + new_scale_color() + layout_circle(HS1_cisplus_end2_gr, geom = "point", aes(shape = gene_label), radius = 8.1, color = col1, trackWidth = 0.3, size = 3) + layout_circle(HS1_cisminus_end2_gr, geom = "point", aes(shape = gene_label), radius = 7.5, color = col2, trackWidth = 0.3, size = 3) + scale_shape_manual(values = c("detox" = 16, "non-detox" = 1)) + layout_circle(HS1_cisplus_end1_gr, geom = "scale", size = 2, radius = 8.5, trackWidth = 0.5) 
# 
pdf("plot/link_HS1wcis.pdf", width = 8, height = 8)
print(link_p)
dev.off()
```

done
