---
title: "R Notebook"
output: html_notebook
---

```{r}
suppressMessages(library("ggplot2"))
suppressMessages(library("reshape2"))
suppressMessages(library("ggprism"))
suppressMessages(library("stringr"))
suppressMessages(library("dplyr"))
```

function to count the number of positions within window
```{r}
# df should have chromosome/position columns; chrlen should have be a dataframe)
sliding_win <- function(df, win, sliding, chrlen) {
  chrs <- unique(df$chromosome)
  step <- ceiling(win*sliding)
  pos_freq_chr <- list()
  for (chr in chrs) {
    sub_df <- df[df$chromosome == chr, ]
    chr_len <- chrlen[chrlen$chromosome == chr, "length"]
    numOfchunks <- ceiling((chr_len-win)/step)
    pos_freq_df <- data.frame(chromosome = chr, bin = seq(0, numOfchunks, 1), win_start = seq(0, numOfchunks, 1), win_end = seq(0, numOfchunks, 1), freq = seq(0, numOfchunks, 1))
    n <- 0
    for (i in seq(0, numOfchunks*step, step)) {
      sub_pos_count <- nrow(sub_df[(sub_df$position >= i) & (sub_df$position < i+win), ])
      pos_freq_df[pos_freq_df$bin == n, "win_start"] <- i
      pos_freq_df[pos_freq_df$bin == n, "win_end"] <- i + win
      pos_freq_df[pos_freq_df$bin == n, "freq"] <- sub_pos_count
      n <- n+1
    }
    pos_freq_chr[[chr]] <- pos_freq_df
  }
  df <- do.call(rbind, pos_freq_chr)
  return(df)
}
```

sample detail of each chromosome recombination (file)
```{r}
### frequency table for each sample on individual chromosome
breakpoint_position <- read.table("data/all_4301.txt", sep = "\t", header = T)
bp <- breakpoint_position[breakpoint_position$left != "-", ]
bp_sample_freq <- table(bp$sample, bp$chromosome) %>% as.data.frame()
colnames(bp_sample_freq) <- c("sample", "chromosome", "freq")
#
bp_sample_freq_chr <- reshape(bp_sample_freq, idvar = "sample", timevar = "chromosome", direction = "wide")
write.table(bp_sample_freq_chr, file = "output/breakpoint_sample_chr.txt", sep = "\t", quote = F, row.name = F)
```

```{r}
breakpoint_position <- read.table("data/all_4301.txt", sep = "\t", header = T)
bp <- breakpoint_position[breakpoint_position$left != "-", ]
bp$distance <- as.numeric(bp$right) - as.numeric(bp$left)
bp50k <- bp[bp$distance < 50000, ]
bp10k <- bp[bp$distance < 10000, ]
nrow(bp50k)/nrow(bp)
nrow(bp10k)/nrow(bp)
```

each chromosome percentage of recombination (data report)
```{r}
# recombination of each sample
bp_sample <- read.table("output/breakpoint_sample_chr.txt", sep= "\t", header = T, row.names = 1)
bp_sample$total <- rowSums(bp_sample)
sample_bp_freq <- as.data.frame(table(bp_sample$total))
colnames(sample_bp_freq) <- c("recombination", "freq")
# 
chr1_freq <- as.data.frame(table(bp_sample$freq.chromosome_1))
chr2_freq <- as.data.frame(table(bp_sample$freq.chromosome_2))
chr3_freq <- as.data.frame(table(bp_sample$freq.chromosome_3))
# recombination total of each chromosome
breakpoint_position <- read.table("data/all_4301.csv", sep = "\t", header = T)
bp <- breakpoint_position[breakpoint_position$left != "-", ]
print(paste0("total recombination ", nrow(bp))) # 
bp_chr1 <- nrow(bp[bp$chromosome == "chromosome_1", ])
bp_chr2 <- nrow(bp[bp$chromosome == "chromosome_2", ])
bp_chr3 <- nrow(bp[bp$chromosome == "chromosome_3", ])
print(paste0("Chr. 1 ", bp_chr1)) # 1109/2927 (37.9%)
print(paste0("Chr. 2 ", bp_chr2)) # 1004/2927 (34.3%)
print(paste0("Chr. 3 ", bp_chr3)) # 814/2927 (27.8%)
```

chromosome detail of frequency in recombination events (file)
```{r}
### frequency table for each chromosome with each possible recombination event counts
breakpoint_sample_freq_chr <- read.table("output/breakpoint_sample_chr.txt", sep= "\t", header = T)
breakpoint_sample_freq_chr_w <- melt(breakpoint_sample_freq_chr, id.vars = "sample")
colnames(breakpoint_sample_freq_chr_w) <- c("sample", "chromosome", "freq")
breakpoint_count_freq <- breakpoint_sample_freq_chr_w %>% group_by(chromosome, freq) %>% summarise(n = n()) %>% as.data.frame()
write.table(breakpoint_count_freq, file = "output/breakpoint_chromosome_freq.txt", sep = "\t", quote = F, row.names = F)
```

bar plot of frequency of recombination events each chromosome (plot)
```{r}
# for Fig. 1b
# bar plot to show the frequency of recombination events on each chromosome
### absolute number of frequency of each recombination event count on each chromosome
freq_table <- read.table("output/breakpoint_chromosome_freq.txt", sep = "\t", header = T)
freq_plot <- ggplot(data = freq_table, aes(x = freq, y = n, fill = chromosome)) + geom_col(position = position_dodge(width = 0.8), width = 0.6, color = "black", size = 0.5) + theme(panel.background = element_rect(color = "grey", fill = "transparent"), panel.grid.major.y = element_line(linetype = "dashed", color = "grey80"), panel.grid.minor.y = element_blank(), panel.grid.major.x = element_blank()) + xlab("recombination events") + ylab("frequency") + scale_fill_manual(values = c("freq.chromosome_1" = "white", "freq.chromosome_2" = "grey",  "freq.chromosome_3" = "black")) + theme(legend.position = "none")
ggsave(plot = freq_plot, file = "plot/recombination_freq_bar.pdf", width = 4.5, height = 4)
```

recombination density across chromosome with window 200 k and step 100 k (file)
```{r}
### report recombination density within some window size and sliding window across chromosome
bp <- read.table("data/all_4301.txt", sep = "\t", header = T)
bp <- bp[bp$left != "-", ]
bbp <- bp[, c("chromosome", "end")]
colnames(bbp) <- c("chromosome", "position")
# 
chrlen <- read.table("data/chrlen.txt", sep = "\t", header = T)
breakpoint200k <- sliding_win(df = bbp, win = 200000, sliding = 0.5, chrlen = chrlen)
write.table(breakpoint200k, file = "output/bp_density_200k(0.5).txt", sep = "\t", quote = F, row.names = F)
```

all/transcribed gene density across chromosome with window 100 k (file)
```{r}
### report gene density of window with sliding size (prepare for both all gene and transcribed genes) using sliding_win
gene_all <- read.table("data/gene_loc_2022.txt", sep = "\t", header = T)
htseq <- read.table("data/htseq_star_new_v2_quantile_90%.txt", sep = "\t", header = T, row.names = 1)
gene_transcribed <- gene_all[gene_all$gene %in% row.names(htseq), ]
gene_all <- gene_all[grepl("chromosome", gene_all$chromosome), ]
gene_transcribed <- gene_transcribed[grepl("chromosome", gene_transcribed$chromosome), ]
chrlen <- read.table("data/chrlen.txt", sep = "\t", header = T)
# 
gene_all <- gene_all[, c("chromosome", "start")]
gene_transcribed <- gene_transcribed[, c("chromosome", "start")]
colnames(gene_all) <- c("chromosome", "position")
colnames(gene_transcribed) <- c("chromosome", "position")
allgene_100k <- sliding_win(gene_all, 100000, 1, chrlen)
transcribedgene_100k <- sliding_win(gene_transcribed, 100000, 1, chrlen)
#
write.table(allgene_100k, file = "output/gene_density_100k.txt", sep = "\t", quote = F, row.names = F)
write.table(transcribedgene_100k, file = "output/transcribed_gene_density_100k.txt", sep = "\t", quote = F, row.names = F)
```

density of SNP of DNA-seq alignment across chromosome with window 200 k and step 100 k (file)
```{r}
### report DNA-SNP density of window with sliding size using sliding_win
DNA_SNP <- read.table("data/eQTL_SNP_gatk4.2.allele.txt", sep = "\t", header = T)
DNA_SNP <- DNA_SNP[grepl("chromosome", DNA_SNP$chromosome), ]
chrlen <- read.table("data/chrlen.txt", sep = "\t", header = T)
DNA_200k <- sliding_win(DNA_SNP, 200000, 0.5, chrlen)
write.table(DNA_200k, file = "output/SNP_density_gatk4.2_200k(0.5).txt", sep = "\t", quote = F, row.names = F)
```

SNP proportion on each chromosome (data report)
```{r}
### report the proportion of SNPs on each chromosome
DNA_SNP_total <- read.table("data/eQTL_SNP_gatk4.2.allele.txt", sep = "\t", header = T) # 724563
chr1_DNA <- DNA_SNP_total[DNA_SNP_total$chromosome == "chromosome_1", ] # 273511/724563 (37.7%)
chr2_DNA <- DNA_SNP_total[DNA_SNP_total$chromosome == "chromosome_2", ] # 246431/724563 (34.0%)
chr3_DNA <- DNA_SNP_total[DNA_SNP_total$chromosome == "chromosome_3", ] # 190068/724563 (26.2%)
```

genotype ratio across chromosome with step 50 k (file)
```{r}
# report the genotype ratio in the population at each step
sample_genotype <- read.table("data/all_4301.txt", sep = "\t", header = T)
chrlen <- read.table("data/chrlen.txt", sep = "\t", header = T)
step_size <- 50000
chrs <- unique(sample_genotype$chromosome)
###
chr_c <- c()
site_c <- c()
homo_c <- c()
hete_c <- c()
for (chr in chrs) {
  sample_chr <- sample_genotype[sample_genotype$chromosome == chr, ]
  length_chr <- chrlen[chrlen$chromosome == chr, "length"]
  for (i in seq(1, length_chr, step_size)) {
    site_genotype <- sample_chr[sample_chr$start <= i & sample_chr$end >= i, "genotype"]
    site_tab <- as.data.frame(table(site_genotype))
    colnames(site_tab) <- c("genotype", "freq")
    site_homo <- site_tab[site_tab$genotype == 0, "freq"]
    site_hete <- site_tab[site_tab$genotype == 1, "freq"]
    chr_c <- c(chr_c, chr)
    site_c <- c(site_c, i)
    homo_c <- c(homo_c, site_homo)
    hete_c <- c(hete_c, site_hete)
  }
}
genotype_count_step <- data.frame(chromosome = chr_c, position = site_c, homozygous = homo_c, heterozygous = hete_c)
write.table(genotype_count_step, file = "output/genotype_ratio_50k.txt", sep = "\t", quote = F, row.names = F)
```

density of recombination and genotype ratio across chromosome (plot)
```{r}
# For figure 1c 
# recombination event frequency and genotype ratio
recombination_freq <- read.table("output/bp_density_200k(0.5).txt", sep = "\t", header = T)
genotype_freq <- read.table("output/genotype_ratio_50k.txt", sep = "\t", header = T)
###
p <- ggplot() + geom_hline(yintercept = 0.5, alpha = 0.2, size = 0.2) + geom_vline(xintercept = c(0, 5e+6, 10e+6, 15e+6, 20e+6, 25e+6, 30e+6), linetype = "dashed", size = 0.2, alpha = 0.2) 
p <- p + geom_area(data = recombination_freq, aes(x = (win_start + win_end)/2, y = freq/30), fill = "#e5855d", alpha = 0.5) + geom_line(data = recombination_freq, aes(x = (win_start + win_end)/2, y = freq/30), color = "#dd6e4c", size = 0.3)
p <- p + geom_line(data = genotype_freq, aes(x = position, y = heterozygous/(homozygous+heterozygous)), color = "#452a3d", size = 0.5) 
p <- p + facet_wrap(~chromosome, ncol = 1, strip.position = "left") + scale_y_continuous(position = "right")
p <- p + xlab("Chromosome position (Mb)") + ylab("percentage of heterozygous genotype (%) & recombination events (30x)")
p <- p + theme(panel.background = element_rect(fill = "transparent", color = "grey"), panel.grid = element_blank())
### + scale_x_continuous(breaks = seq(0, 32654540, 1e+6)) , axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)
ggsave(plot = p, filename = "plot/genotype_ratio_recombination_frequency.pdf", width = 8.5, height = 5.8) 
```

density of (all/transcribed) genes across chromosome using heatmap (plot)
```{r}
gene_freq <- read.table("output/gene_density_100k.txt", sep = "\t", header = T)
transcribed_gene_freq <- read.table("output/transcribed_gene_density_100k.txt", sep = "\t", header = T)
# For figure 1c (gene density as a heatmap: y-axis fixed bar plot)
gene_density <- ggplot(data = gene_freq) + geom_rect(aes(xmin = win_start, xmax = win_end, ymin = 0, ymax = 0.3, fill = freq)) + geom_vline(xintercept = c(0, 5e+6, 10e+6, 15e+6, 20e+6, 25e+6, 30e+6), linetype = "dashed", size = 0.2, alpha = 0.2) + facet_wrap(~chromosome, ncol = 1, strip.position = "left") + scale_fill_viridis_c(direction = -1, option = "C") + theme(legend.position = "bottom", panel.background = element_rect(fill = NA, color = "grey"), panel.grid = element_blank(), axis.ticks.y = element_blank(), axis.text.y = element_blank()) + ylab("") + scale_y_continuous() + theme(legend.position = "bottom")
ggsave(plot = gene_density, filename = "plot/gene_density_100k.pdf", width = 8.5, height = 4)
# 
transcribed_gene_density <- ggplot(data = transcribed_gene_freq) + geom_rect(aes(xmin = win_start, xmax = win_end, ymin = 0, ymax = 0.3, fill = freq)) + geom_vline(xintercept = c(0, 5e+6, 10e+6, 15e+6, 20e+6, 25e+6, 30e+6), linetype = "dashed", size = 0.2, alpha = 0.2) + facet_wrap(~chromosome, ncol = 1, strip.position = "left") + scale_fill_viridis_c(direction = -1, option = "C") + theme(legend.position = "bottom", panel.background = element_rect(fill = NA, color = "grey"), panel.grid = element_blank(), axis.ticks.y = element_blank(), axis.text.y = element_blank()) + ylab("") + scale_y_continuous() + theme(legend.position = "bottom")
ggsave(plot = transcribed_gene_density, filename = "plot/transcribed_gene_density_100k.pdf", width = 8.5, height = 4)
```

density of SNPs of DNA-seq alignment with window 200 k and step 100 k (plot)
```{r}
# supplementary figure
### SNPs of DNA density across chromosomes
SNPs <- read.table("output/SNP_density_gatk4.2_200k(0.5).txt", sep = "\t", header = T)
rect_area <- data.frame(start = c(0, 2000), stop = c(1000, 3000))
SNP_density <- ggplot() + geom_line(data = SNPs, aes(x = win_start, y = freq)) + facet_wrap(~chromosome, nrow = 1, strip.position = "top", scales = "fixed") + theme(panel.background = element_rect(fill = "transparent", color = "grey"), panel.grid.major.x = element_line(linetype = "dashed", color = "grey", size = 0.3), panel.grid.major.y = element_blank(), panel.grid.minor.y = element_blank(), panel.grid.minor.x = element_blank()) + ylab("SNP count per 200 kb window") + scale_y_continuous(position = "left") + xlab("Chromosome position (Mb)") + geom_rect(data = rect_area, aes(ymin = start, ymax = stop, xmin = -Inf, xmax = Inf), fill = "grey", alpha = 0.2)
ggsave(plot = SNP_density, filename = "plot/SNP_density_gatk4.2.pdf", width = 9, height = 2.8)
```

breakpoint density on chromosome with window of 100 k and step 100 k
```{r}
bp <- read.table("data/all_4301.txt", sep = "\t", header = T)
bp <- bp[bp$left != "-", ]
bbp <- bp[, c("chromosome", "end")]
colnames(bbp) <- c("chromosome", "position")
# 
chrlen <- read.table("data/chrlen.txt", sep = "\t", header = T)
breakpoint200k <- sliding_win(df = bbp, win = 100000, sliding = 1, chrlen = chrlen)
write.table(breakpoint200k, file = "output/bp_density_100k.txt", sep = "\t", quote = F, row.names = F)
```

test the association between transcribed gene density and recombination density on chromosome
```{r}
transcribed_gene_freq <- read.table("output/transcribed_gene_density_100k.txt", sep = "\t", header = T)
bp_freq <- read.table("output/bp_density_100k.txt", sep = "\t", header = T)
bp_gene <- merge(bp_freq, transcribed_gene_freq, by = c("bin", "chromosome", "win_start", "win_end"))
colnames(bp_gene) <- c("bin", "chromosome", "win_start", "win_end", "bp_freq", "gene_freq")
bp_p <- ggplot() + geom_point(data = bp_gene, aes(x = bp_freq, y = gene_freq), size = 0.4) + theme(panel.background = element_rect(fill = "transparent", color = "grey"), panel.grid = element_blank()) + xlab("breakpoint frequency (100k)") + ylab("transcribed gene frequency (100k)")
ggsave(plot = bp_p, filename = "plot/breakpoint_gene_freq.pdf", width = 5, height = 5)
```

test the deviation of genotype ratio from Mendelian ratio of 1:1 (chi-square test)
```{r}
geno_ratio <- read.table("output/genotype_ratio_50k.txt", sep = "\t", header = T)
ratio_test <- geno_ratio[, c("homozygous", "heterozygous")]
ratio_test <- cbind(ratio_test, (t(apply(ratio_test, 1, function(x) {
  ch <- chisq.test(x)
  c(unname(ch$statistic), ch$p.value)
}))))
colnames(ratio_test)[3:4] <- c("x-sqaured", "p_value")
geno_ratio_test <- cbind(geno_ratio[, c(1,2)], ratio_test)
geno_ratio_test$padj_BH <- p.adjust(geno_ratio_test$p_value, method = "BH")
geno_ratio_test$padj_bon <- p.adjust(geno_ratio_test$p_value, method = "bonferroni")
write.table(geno_ratio_test, file = "output/genotype_ratio_50k_test.txt", sep = "\t", quote = F, row.names = F)
```

permutation plot of randomly assigned recombination events (50000 times permutations with 1.5 window size)
```{r}
file_plot <- function(file, out) {
  chr <- read.table(file, sep = "\t", header = T)
  obs <- chr["observed", ]
  chr_perm <- chr[-c(1), , drop = F]
  chr_perm_rank <- chr_perm[order(chr_perm$diff, decreasing = T), , drop = F]
  chr_perm_rank$rank <- seq(1, nrow(chr_perm_rank))
  rank_min <- floor(nrow(chr_perm_rank)*0.001)
  rank_limit <- chr_perm_rank[chr_perm_rank$rank == rank_min, "diff"]
  chr_p <- ggplot(data = chr_perm) + geom_histogram(aes(x = diff), binwidth = 2, fill = "#F79044", alpha = 0.4) + geom_vline(xintercept = obs, linetype = "dashed", size = 0.4, color = "red") + geom_vline(xintercept = rank_limit, linetype = "dashed", size = 0.4, color = "blue") + theme_classic() + xlab(out) + ylab("frequency")
  ggsave(plot = chr_p, filename = paste0(out, ".pdf"), width = 8, height = 5)
}
file_plot("permutation/chromosome_1.txt", "chr1_1.5Mb_50000perm")
file_plot("permutation/chromosome_2.txt", "chr2_1.5Mb_50000perm")
file_plot("permutation/chromosome_3.txt", "chr3_1.5Mb_50000perm")
```
