---
title: "R Notebook"
output: html_notebook
---

```{r}
suppressMessages(library("GenomicRanges"))
suppressMessages(library("ggbio"))
suppressMessages(library("ggplot2"))
suppressMessages(library("ggnewscale"))
suppressMessages(library("ggmsa"))
suppressMessages(library("egg"))
suppressMessages(library("stringr"))
suppressMessages(library("dplyr"))
```

Visualize MSA for publication purposes
```{r}
fasta <- "output/NHR_prot.afa"
fasta_df <- read.table(fasta, sep = "\t", header = F)
fasta_df$length <- str_count(fasta_df$V1)
tlen <- max(fasta_df$length)
row_l <- 60
for (i in seq(1, max(tlen), row_l)) {
  p <- ggmsa(fasta, start = i, end = i+row_l-1, seq_name = T, char_width = 0.5, color = "Clustal", font = "helvetical", show.legend = F) + geom_seqlogo(color = "Clustal", top = T)
  pdf(paste0("plot/NHR_", i, ".pdf"), width = 15, height = 3)
  print(p)
  dev.off()
}
```

collect all nuclear receptor genes based on the key word from the gene information file
```{r}
info <- read.table("../../Genome/Annotation/Tetur_gene_info_20190125.txt", sep = "\t", header = T, quote = "")
NHR_info <- info[grepl("nuclear hormone", info$description, ignore.case = T) | grepl("nuclear hormone", info$protein_domains, ignore.case = T) | grepl("nuclear receptor", info$description, ignore.case = T), ]
write.table(NHR_info[, c("Gene_ID", "description", "GO_labels", "protein_domains")], file = "output/NHR_info.txt", sep = "\t", quote = F, row.names = F)
```

compare the geneinfo_based NHR to NHR set from Simon's publication
(Simon gene set include NHR96-like but not other family)
```{r}
simon <- read.table("data/Snoeck_2019_NHR_from_Marilou.txt", sep = "\t", header = F, comment.char = "#")
info <- read.table("output/NHR_info.txt", sep = "\t", header = T)
colnames(simon) <- c("gene", "label")
nrow(simon) # 55
nrow(info) # 97
intersect_NHR <- intersect(info$gene, simon$gene) # 55 
# figure out what genes from info collection but not in simon collection
info_uni <- setdiff(info$gene, simon$gene)
info_sub <- info[info$gene %in% info_uni, ]
write.table(info_sub, "output/NHR_info_notSimon.txt", sep = "\t", quote = F, row.names = F)
info_simon <- info[info$gene %in% simon$gene, ]
write.table(info_simon, "output/NHR_info_Simon.txt", sep = "\t", quote = F, row.names = F)
```

collect NHR locations and eQTL bins with NHR genes within its range
```{r}
loc <- read.table("../data_important/gene_loc_2022.txt", sep = "\t", header = T)
NHR <- read.table("output/NHR_info.txt", sep = "\t", header = T)
NHR_loc <- loc[loc$gene %in% NHR$gene, ]
simon <- read.table("output/NHR_info_Simon.txt", sep = "\t", header = T)
NHR_loc$label <- ifelse(NHR_loc$gene %in% simon$gene, "HR96", "other")
write.table(NHR_loc, "output/NHR_loc.txt", sep = "\t", quote = F, row.names = F)
# 
eQTL <- read.table("../data_important/eQTL_nbin_nhtseq_label_exp.anova.parsed.1e-4.txt", sep = "\t", header = T)
NHR_bin <- c()
for (i in rownames(NHR_loc)) {
  nhr_chr <- NHR_loc[i, "chromosome"]
  nhr_start <- NHR_loc[i, "start"]
  nhr_end <- NHR_loc[i, "end"]
  nhr_bin <- eQTL[eQTL$bin_chromosome == nhr_chr & eQTL$bin_start < (nhr_start+nhr_end)/2 & eQTL$bin_end > (nhr_start+nhr_end)/2, ]$SNP
  NHR_bin <- c(NHR_bin, nhr_bin)
}
NHR_bin <- unique(NHR_bin)
NHR_bin <- as.data.frame(NHR_bin)
write.table(NHR_bin, file = "output/NHR_bin.txt", sep = "\t", quote = F, row.names = F)
```

```{r}
# Collect associations between NHR_bin (extended) and its trans-regulated target genes
eQTL <- read.table("../data_important/eQTL_nbin_nhtseq_label_exp.anova.parsed.1e-4.txt", sep = "\t", header = T)
NHR_bin <- read.table("output/NHR_bin.txt", sep = "\t", header = T)
NHR_b <- c(NHR_bin$NHR_bin, "chromosome_3:15490119-15674051", "chromosome_1:20863487-20951636")
eQTL_bin <- eQTL[eQTL$SNP %in% NHR_b, ]
eQTL_pos <- unique(eQTL_bin[, c("bin_chromosome", "bin_position")])
# extend bin range
NHR_extend <- c()
for (i in rownames(eQTL_pos)) {
  echr <- eQTL_pos[i, "bin_chromosome"]
  epos <- eQTL_pos[i, "bin_position"]
  esnp <- eQTL[eQTL$bin_chromosome == echr & abs(epos - eQTL$bin_position) < 100e+3, "SNP"]
  NHR_extend <- c(NHR_extend, esnp)
}
# 
eQTL_nhr <- eQTL[eQTL$SNP %in% unique(NHR_extend) & eQTL$label == "trans" & eQTL$gene_label == "detox", ]
write.table(eQTL_nhr, file = "output/NHR_trans.txt", sep = "\t", quote = F, row.names = F)
```

plotting circle chromosomes for the NHR-trans
```{r}
chr <- read.table("../data_important/chrlen.txt", sep = "\t", header = T)
NHR_loc <- read.table("output/NHR_loc.txt", sep = "\t", header = T)
eQTL_nhr <- read.table("output/NHR_trans.txt", sep = "\t", header = T)
simon <- read.table("output/NHR_info_Simon.txt", sep = "\t", header = T)
NHR_loc$label <- ifelse(NHR_loc$gene %in% simon$gene, "HR96", "other")
NHR_loc_chr <- NHR_loc[grepl("chromosome", NHR_loc$chromosome), ]
NHR_loc_chr <- NHR_loc_chr[, c("chromosome", "start", "end", "gene", "label")]
# 
eQTL_bin <- eQTL_nhr[, c("bin_chromosome", "bin_start", "bin_end", "beta")]
eQTL_bin$effect <- ifelse(eQTL_bin$beta > 0, "+", "-")
eQTL_gene <- eQTL_nhr[, c("gene_chromosome", "gene_start", "gene_end", "gene")]
colnames(eQTL_bin) <- c("chromosome", "start", "end", "beta", "effect")
colnames(eQTL_gene) <- c("chromosome", "start", "end", "gene")
# 
bin.gr <- makeGRangesFromDataFrame(eQTL_bin, keep.extra.columns = T, ignore.strand = T)
gene.gr <- makeGRangesFromDataFrame(eQTL_gene, keep.extra.columns = T, ignore.strand = T)
nhr.gr <- makeGRangesFromDataFrame(NHR_loc_chr, keep.extra.columns = T, ignore.strand = T)
seqlengths(bin.gr) <- chr$length
seqlengths(gene.gr) <- chr$length
seqlengths(nhr.gr) <- chr$length
bin.gr$to.gr <- gene.gr
# 
nhr_p <- ggplot() + layout_circle(bin.gr, geom = "link", linked.to = "to.gr", aes(color = effect), size = 0.4, radius = 8, trackWidth = 0.5) + scale_color_manual(values = c("+" = "black", "-" = "grey")) + layout_circle(bin.gr, geom = "ideo", fill = "grey", alpha = 0.4, radius = 8.1, trackWidth = 0.8, color = "black") + layout_circle(bin.gr, geom = "scale", size = 1, radius = 8.9, trackWidth = 0.5) + new_scale_color() + layout_circle(nhr.gr, geom = "point", shape = 16, aes(color = label),  radius = 6) + scale_color_manual(values = c("HR96" = "red", "other" = "orange")) + layout_circle(gene.gr, geom = "point", shape = 23, radius = 5.8)
pdf("plot/NHR_circlular.pdf", width = 10, height = 8)
nhr_p
dev.off()
```

detox genes with trans-eQTLs and the NHR location
```{r}
NHR_loc <- read.table("output/NHR_loc.txt", sep = "\t", header = T)
detox <- read.table("../data_important/detox_RC.txt", sep = "\t", header = T)
eQTL <- read.table("../data_important/eQTL_nbin_nhtseq_label_exp.anova.parsed.1e-4.txt", sep = "\t", header = T)
eQTL <- eQTL[eQTL$p.value < 1e-10, ]
detox_trans <- eQTL[eQTL$label == "trans" & eQTL$gene_label == "detox", ]
detox_trans.b <- detox_trans[, c("bin_chromosome", "bin_start", "bin_end", "beta")]
detox_trans.b$effect <- ifelse(detox_trans.b$beta > 0, "+", "-")
detox_trans.g <- detox_trans[, c("gene_chromosome", "gene_start", "gene_end", "gene")]
colnames(detox_trans.b) <- c("chromosome", "start", "end", "beta", "effect")
colnames(detox_trans.g) <- c("chromosome", "start", "end", "gene")
NHR_loc <- NHR_loc[grepl("chromosome", NHR_loc$chromosome), c("chromosome", "start", "end", "label", "gene")]
# 
chr <- read.table("../data_important/chrlen.txt", sep = "\t", header = T)
bin.gr <- makeGRangesFromDataFrame(detox_trans.b, keep.extra.columns = T, ignore.strand = T)
gene.gr <- makeGRangesFromDataFrame(detox_trans.g, keep.extra.columns = T, ignore.strand = T)
nhr.gr <- makeGRangesFromDataFrame(NHR_loc, keep.extra.columns = T, ignore.strand = T)
seqlengths(bin.gr) <- chr$length
seqlengths(gene.gr) <- chr$length
seqlengths(nhr.gr) <- chr$length
bin.gr$to.gr <- gene.gr
# 
p <- ggplot() + layout_circle(bin.gr, geom = "link", linked.to = "to.gr", aes(color = effect), size = 0.4, radius = 7.7, trackWidth = 0.5) + scale_color_manual(values = c("+" = "black", "-" = "grey")) + layout_circle(bin.gr, geom = "ideo", fill = "grey", alpha = 0.4, radius = 8.1, trackWidth = 0.8, color = "black") + layout_circle(bin.gr, geom = "scale", size = 1, radius = 8.9, trackWidth = 0.5) + new_scale_color() + layout_circle(nhr.gr, geom = "rect", aes(color = label),  radius = 8, trackWidth = 0.5)  + layout_circle(gene.gr, geom = "point", shape = 21, radius = 5.5)
# pdf("plot/NHR_circlular.pdf", width = 10, height = 8)
```

compare two HR96 gene copies expression
```{r}

```

add gene expression for HR96 copies 
```{r}
bar_norm_exp <- function(norm_file, sample_1, sample_2, out) {
  norm_exp <- read.table(norm_file, sep = "\t", header = T)
  #
  sample_1_exp <- norm_exp[c("HR96_copy1", "HR96_copy2"), grepl(sample_1, colnames(norm_exp))]
  s1 <- t(sample_1_exp)
  sample_2_exp <- norm_exp[c("HR96_copy1", "HR96_copy2"), grepl(sample_2, colnames(norm_exp))]
  s2 <- t(sample_2_exp)
  s1_t <- t.test(s1[, "HR96_copy1"], s1[, "HR96_copy2"], paired = T, alternative = "two.sided")
  s2_t <- t.test(s2[, "HR96_copy1"], s2[, "HR96_copy2"], paired = T, alternative = "two.sided")
  print(paste(sample_1, "estimate", s1_t$estimate, "p-value", s1_t$p.value))
  print(paste(sample_2, "estimate", s2_t$estimate, "p-value", s2_t$p.value))
  #
  HR96 <- norm_exp[c("HR96_copy1", "HR96_copy2"), ]
  HR96 <- reshape2::melt(as.matrix(HR96))
  colnames(HR96) <- c("gene", "sample", "expression")
  HR96$strain <- gsub("_[1-5|A-E]$", "", HR96$sample)
  HR96$strain <- factor(HR96$strain, levels = c(sample_1, sample_2))
  HR96_summ <- HR96 %>% group_by(strain, gene) %>% summarise(sd=sd(expression), expression=mean(expression), n=n()) %>% as.data.frame()
  HR96_summ$sem <- HR96_summ$sd/(HR96_summ$n)^0.5
  # 
  HR96_inbred <- ggplot(data = HR96, aes(x = gene, y = expression, fill = strain)) + geom_bar(data = HR96_summ, position = position_dodge(width = 0.8), width = 0.7, stat = "identity", color = "black") + theme_classic() + geom_errorbar(data = HR96_summ, aes(ymin = expression-2*sem, ymax = expression+2*sem), width = 0.2, size = 0.4, position=position_dodge(width = 0.8)) + scale_fill_manual(values = c("black", "white")) + geom_jitter(position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.8), color = "grey", alpha = 0.7) + xlab("") + theme(legend.position = "right")
  pdf(paste0("plot/", out, ".pdf"), width = 4, height = 3)
  print(HR96_inbred)
  dev.off()
}
```

```{r}
bar_norm_exp("data/htseq/inbred_norm.txt", "MRVPi", "ROS_ITi", "inbred")
bar_norm_exp("data/htseq/A_NIL_norm.txt", "A18_E", "A18_H", "A_NIL")
bar_norm_exp("data/htseq/C_NIL_norm.txt", "C31_E", "C31_H", "C_NIL")
bar_norm_exp("data/htseq/RNAi_bean_norm.txt", "NHR", "GFP", "RNAi_bean")
bar_norm_exp("data/htseq/RNAi_tomato_norm.txt", "TN", "TB", "RNAi_tomato")
```

