---
title: "R Notebook"
output: html_notebook
---

```{r}
suppressMessages(library("ggplot2"))
suppressMessages(library("dplyr"))
suppressMessages(library("ggrepel"))
suppressMessages(library("VennDiagram"))
suppressMessages(library("ComplexHeatmap"))
suppressMessages(library("viridis"))
suppressMessages(library("circlize")) # for circlize colors
suppressMessages(library("dendsort"))
```

draw gene model within bin region (peak bin selected)
```{r}
# After write out exon position, I manually selected gene exons for plotting
# from 12490000 - 12525000
bin_exon <- read.table("data/bin_peak_exon.txt", sep = "\t", header = T)
model_start <- min(bin_exon$start)%/%5000*5000
model_end <- (max(bin_exon$end)%/%5000+1)*5000
whole_length <- model_end - model_start
pdf("plot/chr1_12490-12530kb.pdf", width = 15, height = 3)
options(repr.plot.width=15, repr.plot.height=3)
plot.new()
segments(0, 0.8, 1, 0.8, lwd = 0.5)
### draw the x-axis scale for gene position
for (s in seq(model_start, model_end, 5000)) {
  dot <- (s - model_start)/whole_length
  segments(dot, 0.8-0.02, dot, 0.8+0.02, lwd = 1)
}
# add exon information
plus <- 0.5
minus <- 0.2
for (r in row.names(bin_exon)) {
  stt <- bin_exon[r, "start"]
  end <- bin_exon[r, "end"]
  strand <- bin_exon[r, "strand"]
  left <- (stt - model_start)/whole_length
  right <- (end - model_start)/whole_length
  if (strand == "+") polygon(c(left, left, right, right), c(plus-0.1, plus-0.05, plus-0.05, plus-0.1), col = "black", border = NA)
  else polygon(c(left, left, right, right), c(minus-0.1, minus-0.05, minus-0.05, minus-0.1), col = "black", border = NA)
}
# add gene information
bgid <- unique(bin_exon$gene)
bin_gene <- data.frame(gene = bgid, start = 0, end = 0)
for (g in bgid) {
  bmin <- min(bin_exon[bin_exon$gene == g, ]$start)
  bmax <- max(bin_exon[bin_exon$gene == g, ]$end)
  bin_gene[bin_gene$gene == g, "start"] <- bmin
  bin_gene[bin_gene$gene == g, "end"] <- bmax
}
for (r in row.names(bin_gene)) {
  stt <- bin_gene[r, "start"]
  end <- bin_gene[r, "end"]
  left <- (stt - model_start)/whole_length
  right <- (end - model_start)/whole_length
  segments(left, 0.05, right, 0.05, col = "red")
}
dev.off()
```

filter DEGs for bean and tomota and write out in files (padj < 0.01, baseMean > 10)
```{r}
bean <- read.table("output/NHRvsGFP_res0.01_HR96.txt", sep = "\t", header = T, row.names = 1)
tomato <- read.table("output/TNvsTB_res0.01_HR96.txt", sep = "\t", header = T, row.names = 1)
bean_sig <- row.names(bean[complete.cases(bean) & bean$padj < 0.01, ])
tomato_sig <- row.names(tomato[complete.cases(tomato) & tomato$padj < 0.01, ])
#
all_sig <- union(bean_sig, tomato_sig)
intersect_sig <- intersect(bean_sig, tomato_sig)
# 
bean_sub <- bean[all_sig, c("baseMean", "log2FoldChange", "padj")]
colnames(bean_sub) <- c("bean_bm", "bean_lfc", "bean_padj")
tomato_sub <- tomato[all_sig, c("baseMean", "log2FoldChange", "padj")]
colnames(tomato_sub) <- c("tomato_bm", "tomato_lfc", "tomato_padj")
bean_tomato <- cbind(bean_sub, tomato_sub)
bean_tomato$label <- ifelse(rownames(bean_tomato) %in% intersect_sig, "both", ifelse(rownames(bean_tomato) %in% bean_sig, "bean", "tomato"))
#
write.table(bean_tomato, file = "output/bean_tomato_sig.txt", sep = "\t", quote = F, row.names = T)
write.table(bean[bean_sig, ], file = "output/bean_sig.txt", sep = "\t", quote = F, row.names = T)
write.table(tomato[tomato_sig, ], file = "output/tomato_sig.txt", sep = "\t", quote = F, row.names = T)
```

venn diagram for DEGs (detox) in bean and tomato RNAi (venn plot)
```{r}
bean_sig <- read.table("output/bean_sig.txt", sep = "\t", header = T, row.names = 1)
tomato_sig <- read.table("output/tomato_sig.txt", sep = "\t", header = T, row.names = 1)
detox <- read.table("../data_important/detox_RC.txt", sep = "\t", header = T, row.names = 1)
RNAi <- list()
RNAi[["bean"]] <- row.names(bean_sig)
RNAi[["tomato"]] <- row.names(tomato_sig)
rnai_bean_tomato <- venn.diagram(RNAi, category.names = names(RNAi), filename = NULL, margin = 0.05)
ggsave(plot = rnai_bean_tomato, filename = "plot/bean_tomato_venn.pdf", width = 5, height = 5)
# for detox venn
RNAi_detox <- list()
RNAi_detox[["bean"]] <- row.names(bean_sig)[row.names(bean_sig) %in% row.names(detox)]
RNAi_detox[["tomato"]] <- row.names(tomato_sig)[row.names(tomato_sig) %in% row.names(detox)]
detox_rnai_bean_tomato <- venn.diagram(RNAi_detox, category.names = names(RNAi_detox), filename = NULL, margin = 0.05)
ggsave(plot = detox_rnai_bean_tomato, filename = "plot/detox_bean_tomato_venn.pdf", width = 5, height = 5)
```

Scatter plot for log2FC on bean (x) and on tomato (y) by color coded 
```{r}
detox <- read.table("data/detox_label.txt", sep = "\t", header = T, row.names = 1)
eQTL <- read.table("../data_important/hotspot1.txt", sep = "\t", header = 1)
RNAi <- read.table("output/bean_tomato_sig.txt", sep = "\t", header = T, row.names = 1)
RNAi$detox <- ifelse(rownames(RNAi) %in% rownames(detox), "detox", "non-detox")
detox_HR96 <- rbind(detox[, "name", drop = F], data.frame(row.names = c("HR96_copy1", "HR96_copy2"), name = c("HR96_1a", "HR96_1b")))
RNAi_name <- merge(RNAi, detox_HR96, by = 0, all.x = T)
# write.table(RNAi_name, file = "RNAi.temp", sep = "\t", quote = F, row.names = F)
RNAi_name <- RNAi_name[order(RNAi_name$detox, decreasing = T), ]
RNAi_name$color <- paste0(RNAi_name$detox, "-", RNAi_name$label)
RNAi_name$shape <- ifelse(RNAi_name$Row.names %in% eQTL$gene, "call", "not-call")   
RNAi_p <- ggplot() + geom_point(data = RNAi_name, aes(x = bean_lfc, y = tomato_lfc, color = color, fill = color, alpha = detox, shape = shape), size = 2.1, stroke = 0.3) + scale_shape_manual(values = c("call" = 21, "not-call" = 23)) + scale_color_manual(values = c("detox-both" = "black", "detox-tomato" = "black", "detox-bean" = "black", "non-detox-both" = "#F17020", "non-detox-tomato" = "#F7D340", "non-detox-bean" = "#5C126E"))  + scale_fill_manual(values = c("non-detox-both" = "white", "non-detox-tomato" = "white", "non-detox-bean" = "white", "detox-both" = "#F17020", "detox-tomato" = "#F7D340", "detox-bean" = "#5C126E")) + scale_alpha_manual(values = c("detox" = 0.85, "non-detox" = 1)) + theme(panel.background = element_rect(fill = NA, color = "grey"), panel.grid = element_blank()) + xlab("log2FC bean") + ylab("log2FC tomato") + geom_hline(yintercept = 0, size = 0.5, linetype = "dashed", color  = "grey", alpha = 0.3) + geom_vline(xintercept = 0, size = 0.5, linetype = "dashed", color  = "grey", alpha = 0.3) + geom_abline(intercept = 0, slope = 1, linetype = "dashed", color  = "grey", alpha = 0.3) + coord_fixed(ratio = 1) 
# 
RNAi_p_all <- RNAi_p + geom_text_repel(data = RNAi_name, aes(x = bean_lfc, y = tomato_lfc, label = name), size = 2, fontface = "italic", min.segment.length = 0, segment.size = 0.2, max.overlaps = 80) + theme(legend.position = "right")
RNAi_p_all_2 <- RNAi_p + geom_text_repel(data = RNAi_name, aes(x = bean_lfc, y = tomato_lfc, label = Row.names), size = 0.6, min.segment.length = 0, segment.size = 0.2, max.overlaps = 80) +  theme(legend.position = "right")
#RNAi_name_sub <- RNAi_name[grepl("cyp|ugt|cce|GST|ABC", RNAi_name$name, ignore.case = T), ]
#RNAi_p_sub <- RNAi_p + geom_text_repel(data = RNAi_name_sub, aes(x = bean_lfc, y = tomato_lfc, label = name), size = 2, fontface = "italic", min.segment.length = 0, segment.size = 0.2, max.overlaps = 80) + theme(legend.position = "right")
ggsave(RNAi_p_all, filename = "plot/RNAi_point_label.pdf", width = 6, height = 5)
ggsave(RNAi_p_all_2, filename = "plot/RNAi_point_label_2.pdf", width = 6, height = 5)
ggsave(RNAi_p, filename = "plot/RNAi_point.pdf", width = 6, height = 5)
```

check gene set from manual classification
```{r}
gset <- read.table("output/RNAi_gene_set.txt", sep = "\t", header = T)
RNAi <- read.table("output/RNAi_labeled.txt", sep = "\t", header = T, row.names = 1)
RNAi_set <- list() 
RNAi_set[["set1"]] <- gset[gset$set == "set1", "gene"]
RNAi_set[["set2"]] <- gset[gset$set == "set2", "gene"]
RNAi_set[["other"]] <- gset[gset$set == "other", "gene"]
set1 <- RNAi[row.names(RNAi) %in% RNAi_set[["set1"]], ]
set2 <- RNAi[row.names(RNAi) %in% RNAi_set[["set2"]], ]
other <- RNAi[row.names(RNAi) %in% RNAi_set[["other"]], ]
set1_p <- ggplot(data = set1, aes(x = bean_lfc, y = tomato_lfc)) + geom_point()
set2_p <- ggplot(data = set2, aes(x = bean_lfc, y = tomato_lfc)) + geom_point()
other_p <- ggplot(data = other, aes(x = bean_lfc, y = tomato_lfc)) + geom_point()
# ggsave(plot = set1_p, filename = "plot/temp_set1.pdf", width = 6, height = 5)
# ggsave(plot = set2_p, filename = "plot/temp_set2.pdf", width = 6, height = 5)
# ggsave(plot = other_p, filename = "plot/temp_other.pdf", width = 6, height = 5)
```

z-score of RNAi samples
```{r}
bean_norm <- read.table("data/RNAi_bean_norm.txt", sep = "\t", row.names = 1, header = T)
tomato_norm <- read.table("data/RNAi_tomato_norm.txt", sep = "\t", row.names = 1, header = T)
all_norm <- cbind(bean_norm, tomato_norm[rownames(bean_norm), ])
# write.table(all_norm, file = "output/RNAi_norm.txt", sep = "\t", quote = F, row.names = T)
#
normalized_count_z <- t(apply(all_norm, 1, scale))
colnames(normalized_count_z) <- colnames(all_norm)
write.table(normalized_count_z, file = "output/RNAi_norm_z.txt", sep = "\t", quote = F, row.names = T)
```

```{r}
norm_exp <- read.table("output/RNAi_norm.txt", sep = "\t", header = T)
gset <- read.table("output/RNAi_gene_set.txt", sep = "\t", header = T)
gset <- gset[gset$set != "other", ]
GFP <- read.table("output/TBvsGFP_res0.01_HR96.txt", sep = "\t", header = T, row.names = 1) 
gset_1 <- merge(gset, norm_exp, by.x = "gene", by.y = 0, all.x = T)
gset_2 <- merge(gset_1, GFP, by.x = "gene", by.y = 0, all.x = T)
write.table(gset_2, file = "gset2.txt", sep = "\t", quote = F, row.names = F)
```

heatmap of gene set1 and set2 (with gene name) and add log2FC as bar plot
```{r}
gset <- read.table("output/RNAi_gene_set.txt", sep = "\t", header = T)
gset <- gset[gset$set != "other", ]
gset <- gset[order(gset$set), ]
rownames(gset) <- seq(1, nrow(gset))
z_score <- read.table("output/RNAi_norm_z.txt", sep = "\t", header = T, row.names = 1)
GFP <- read.table("output/TBvsGFP_res0.01_HR96.txt", sep = "\t", header = T, row.names = 1) 
GFP_df <- GFP[GFP$padj < 0.01 & complete.cases(GFP), ] # DEGs for GFP on bean and tomato
detox <- read.table("data/detox_label.txt", sep = "\t", header = T, row.names = 1)
# 
gset_z <- z_score[gset$gene, ] # gset z-score
gset_label <- gset$set # gset label (set1/set2)
gset$sig <- ifelse(gset$gene %in% rownames(GFP_df), "1", "")
# 

gset_sub <- gset[gset$sig == "1", ]
at_loc <- as.numeric(rownames(gset_sub))
# bar of log2FC of genes
GFP_lfc <- GFP[rownames(gset_z), "log2FoldChange"]
gset_detox <- rowAnnotation(rbar = anno_barplot(GFP_lfc, bar_width = 1, gp = gpar(col = "black", fill = "#FDE725"), border = T, width = unit(3, "cm")), show_annotation_name = F) # gset mark annotation for detox
# 
gset_sig <- rowAnnotation(significance = anno_mark(at = at_loc, labels = gset_sub$sig, which = "column", side = "left", labels_gp = gpar(fontsize = 5))) # gset mark annotation for significance
# gset$detox <- ifelse(gset$gene %in% rownames(detox), "*", "")
# gset_sub <- gset[gset$detox == "*", ]
# at_loc2 <- as.numeric(rownames(gset_sub))
col_fun <- colorRamp2(c(-3, 0, 3), c("blue", "white", "red")) # color
gset_p <- Heatmap(as.matrix(gset_z), col = col_fun, column_split = rep(c("A", "B"), c(10, 8)), row_split = gset_label, cluster_columns = F, cluster_rows = T, name = "z-score", show_row_names = T, show_column_names = T, column_names_side = "bottom", show_column_dend = F, show_row_dend = F,  rect_gp = gpar(col = "white", lwd = 0.1), border = "black", right_annotation = gset_detox, left_annotation = gset_sig, heatmap_legend_param = list(direction = "horizontal", at = c(-3, -2, -1, 0, 1, 2, 3), labels = c("-3", "-2", "-1", "0", "1", "2", "3")))
pdf("plot/RNAi_heatmap_bar_HR96_2.pdf", width = 7, height = 12)
draw(gset_p, heatmap_legend_side = "bottom")
dev.off()
pdf("plot/RNAi_heatmap_bar_HR96.pdf", width = 7, height = 10)
draw(gset_p, heatmap_legend_side = "bottom")
dev.off()
```

GO enrichment in bubble plot with gene set on x-axis, and go-term on y-axis (plot)
```{r}
set1_mf <- read.table("output/set1_GO.mf.txt", sep = "\t", header = T, stringsAsFactors = F)
set1_bp <- read.table("output/set1_GO.bp.txt", sep = "\t", header = T, stringsAsFactors = F)
set2_mf <- read.table("output/set2_GO.mf.txt", sep = "\t", header = T, stringsAsFactors = F)
set2_bp <- read.table("output/set2_GO.bp.txt", sep = "\t", header = T, stringsAsFactors = F)
set1_mf$set <- "set1"
set1_bp$set <- "set1"
set2_mf$set <- "set2"
set2_bp$set <- "set2"
set1_mf$term <- "mf"
set1_bp$term <- "bp"
set2_mf$term <- "mf"
set2_bp$term <- "bp"
go_all <- rbind(set1_mf, set1_bp, set2_mf, set2_bp)
go_all <- go_all[with(go_all, order(go_all$set, go_all$term)), ]
write.table(go_all, file = "output/set_goall.txt", sep = "\t", quote = F, row.names = F)
go_all$gnum <- as.numeric(gsub("/.*", "", go_all$GeneRatio))
go_all$bgnum <- as.numeric(gsub(".*/", "", go_all$GeneRatio))
go_all$gprop <- go_all$gnum/go_all$bgnum*100
# 
go_all <- go_all %>% arrange(term, -log10(p.adjust)) %>% mutate(Description = factor(Description, levels = unique(Description)))
# make bubble plot to show 
go_p <- ggplot() + geom_point(data = go_all, aes(x = Description, y = set, color = -log10(p.adjust), size = gprop)) + xlab("") + ylab("") + scale_x_discrete(position = "top") + scale_size(range = c(1, 8), name = "Gene ratio (%)") + theme(panel.background = element_rect(color = "grey", fill = "transparent"), panel.grid = element_blank()) + scale_color_viridis_c(alpha = 0.9, option = "B", direction = -1) + coord_flip()
pdf("plot/GO_bubble.pdf", width = 8.5, height = 5)
print(go_p)
dev.off()
```

host plant of same treatment for differential expression analysis
```{r}
# NHR <- read.table("output/TNvsNHR_res0.01.txt", sep = "\t", header = T, row.names = 1)
GFP <- read.table("output/TBvsGFP_res0.01.txt", sep = "\t", header = T, row.names = 1)
# NHR_d <- NHR[NHR$padj < 0.01 & abs(NHR$log2FoldChange) > 1 & complete.cases(NHR), ] # 755 (withou lfc cutoff: 3688)
GFP_d <- GFP[GFP$padj < 0.01 & abs(GFP$log2FoldChange) > 1 & complete.cases(GFP), ] # 863 (without lfc cutoff: 4236)
set1 <- read.table("output/set1_gid.txt", sep = "\t", header = T, row.names = 1)
set2 <- read.table("output/set2_gid.txt", sep = "\t", header = T, row.names = 1)
# set1_NHR_d <- intersect(rownames(set1), rownames(NHR_d))
set1_GFP_d <- intersect(rownames(set1), rownames(GFP_d))
# set2_NHR_d <- intersect(rownames(set2), rownames(NHR_d))
set2_GFP_d <- intersect(rownames(set2), rownames(GFP_d))
# 
# length(set1_NHR_d)
length(set1_GFP_d)
# length(set2_NHR_d)
length(set2_GFP_d)
```

done

